### PR TITLE
Added move to bottom in layertreeview context menu

### LINKED ISSUE #35530: Crash when Moving to top both a layer and it's parent group
**Describe the bug**
If one selects a layer and a group it belongs to and select *Move to top* from the context menu QGIS may crash depending on the order of selection. If the group was selected first, or a multi selection with *shift* was made, then the group is moved before it's contents and QGIS crashes.

**How to Reproduce**
- Select a group.
- ctrl-click on one of it's layers
- right click and select *Move to top*

**QGIS and OS versions**
any

### PR DESCRIPTION
**Pull request by uclaros**
## Description
Moving around layers on big projects was made easier with *Move to top* but still one had to move manually his basemaps to the bottom of the layer list, a cumbersome task in big layer trees. This PR adds a *Move to bottom* option on the layer tree context menu that works in the same way as Move to top.

Also fixes #35530

### CODE REVIEW

File: src/core/layertree/qgslayertreenode.h [hunk 1]
```diff
@@ -208,6 +208,12 @@ class CORE_EXPORT QgsLayerTreeNode : public QObject
      */
     QList< QgsMapLayer * > checkedLayers() const;
 
+    /**
+     * Returns the depth of this node, i.e. the number of it's ancestors
```
Comment (m-kuhn) [inline src/core/layertree/qgslayertreenode.h:212]:
```suggestion
     * Returns the depth of this node, i.e. the number of its ancestors
```

File: src/gui/layertree/qgslayertreeviewdefaultactions.cpp [hunk 2]
```diff
@@ -429,27 +436,40 @@ void QgsLayerTreeViewDefaultActions::moveOutOfGroup()
...
-  int insertIdx;
-  const QList< QgsLayerTreeNode * >  selectedNodes = mView->selectedNodes();
-  for ( QgsLayerTreeNode *n : selectedNodes )
+  QList< QgsLayerTreeNode * >  selectedNodes = mView->selectedNodes();
+  // sort the nodes by depth first to avoid moving a group before it's contents
+  std::stable_sort( selectedNodes.begin(), selectedNodes.end(), []( const QgsLayerTreeNode * a, const QgsLayerTreeNode * b )
+  {
+    return a->depth() < b->depth();
+  } );
+  for ( auto it = selectedNodes.rbegin(); it < selectedNodes.rend(); ++it )
```
Comment (m-kuhn) [inline src/gui/layertree/qgslayertreeviewdefaultactions.cpp:445]: Can we use an (easier to read) range-based for loop here?

**Review by m-kuhn — COMMENTED** [review]
Nice one!

File: src/gui/layertree/qgslayertreeviewdefaultactions.cpp (hunk 2 above)
Comment (uclaros) [reply src/gui/layertree/qgslayertreeviewdefaultactions.cpp:445]: We need to iterate in reverse in order to preserve the selection order. If we use a range based for, we'll need to have the list reversed first.

File: src/core/layertree/qgslayertreenode.h (hunk 1 above)
Comment (uclaros) [reply src/core/layertree/qgslayertreenode.h:212]: Thanks!

**Comment by uclaros** [comment]
I could not find tests for `QgsLayerTreeNode`. Should I add somewhere a test for the new method `QgsLayerTreeNode::depth()`?

**Review by m-kuhn — APPROVED** [review]

**Comment by m-kuhn** [comment]
> I could not find tests for `QgsLayerTreeNode`. Should I add somewhere a test for the new method `QgsLayerTreeNode::depth()`?

There is tests/src/core/testqgslayertree.cpp . If it doesn't fit in there add a new one :+1:

**Comment by uclaros** [comment]
While it is the closest match, I don't think it fits, no `QgsLayerTreeNode` methods are tested there.

**Comment by m-kuhn** [comment]
In this case just add a new file (in there as cpp test or if you prefer you can also do it on python side, your call).

File: python/core/auto_generated/layertree/qgslayertreenode.sip.in [hunk 3]
```diff
@@ -208,6 +208,13 @@ Returns a list of any checked layers which belong to this node or its
 children.
 
 .. versionadded:: 3.0
+%End
+
+    int depth() const;
+%Docstring
+Returns the depth of this node, i.e. the number of it's ancestors
```
Comment (nyalldawson) [inline python/core/auto_generated/layertree/qgslayertreenode.sip.in:215]: HMm - sipify test should be rejecting this as out-of-sync...

Comment (nyalldawson) [reply python/core/auto_generated/layertree/qgslayertreenode.sip.in:215]: Ahh I see, I don't think you've pushed your "it's" fix yet?

Comment (uclaros) [reply python/core/auto_generated/layertree/qgslayertreenode.sip.in]: I missed that one, should be ok now

**Comment by pigreco** [comment]
Please add `GUI/UX` label

**Comment by m-kuhn** [comment]
@pigreco please apply for labelling power here: https://github.com/orgs/qgis/teams/community/members

**Pull request merged**